hack/licenser/main.go

```diff
@@ -0,0 +1,137 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/template"
+)
+
+var (
+	depFile = flag.String("f", "go.mod", "File with the list of dependencies")
+	dir     = flag.String("d", "", "Project directory")
+)
+
+type Dependency struct {
+	Name    string
+	Version string
+	License string
+}
+
+type Dependencies struct {
+	List []*Dependency
+}
+
+func main() {
+	flag.Parse()
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	deps, err := loadFile(filepath.Join(*dir, *depFile))
+	if err != nil {
+		log.Fatalf("Can't open file with dependencies: %s", err.Error())
+	}
+
+	issues := checkForLicense(deps)
+	if issues > 0 {
+		log.Fatal("Can't create NOTICE.txt, there are issues with dependencies!")
+	}
+
+	createNoticeFile(deps)
+}
+
+func loadFile(path string) (*Dependencies, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var lines []string
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		lines = append(lines, scanner.Text())
+	}
+
+	deps := &Dependencies{}
+	for i, v := range lines {
+		if i > 4 && i < (len(lines)-1) {
+			dep := strings.Split(v, " ")
+			path := dep[0]
+			path = strings.Replace(path, "	", "", -1)
+			deps.List = append(deps.List, &Dependency{Name: path, Version: dep[1]})
+		}
+	}
+
+	return deps, nil
+}
+
+func checkForLicense(deps *Dependencies) int {
+	var issues []string
+	licenses := []string{"LICENSE", "LICENSE.txt", "LICENCE"} // Used to keep all possible names of license files
+	for _, dep := range deps.List {
+		counter := len(licenses)
+		for _, v := range licenses {
+			bytes, err := ioutil.ReadFile(filepath.Join(*dir, "vendor", dep.Name, v))
```

Comment:
I don't think we can count on the vendor directory since the default for go modules is to not use it. Looks like the default src cache is `$GOPATH/pkg/mod`: https://golang.org/cmd/go/#hdr-GOPATH_and_Modules

Reply:
Yeah, we can't count on it out of the box. But I'm thinking about using `go mod vendor` to collect all deps into `vendor` folder. In this case we can rely on content of `vendor` dir.